Initial setup for a GitHub Apps

TrueBrain opened the pull request:
It processes integration events, and knowns of the existing integrations on startup

TrueBrain committed:
Initial setup for DorpsGek GitHub App

It processes integration events, and knowns of the existing integrations on startup

LordAro commented on dorpsgek/__main__.py:
```diff
@@ -0,0 +1,38 @@
+    app.router.add_post("/", github_handler)
+    app.on_startup.append(github_startup)
+
+    web.run_app(app, port=3000)
```
be nice if the port was configurable

LordAro commented on dorpsgek/github.py:
```diff
@@ -0,0 +1,184 @@
+# HACK - Monkeypatch a fix that is already upstream
+@classmethod
+def ratelimit_from_http(cls, headers):
+    try:
```
should also catch ValueError in the case of non-numeric headers

TrueBrain replied:
This is a direct copy of what is currently upstream. I guess that is better than a modification to it? :)

LordAro commented on dorpsgek/github.py:
```diff
@@ -0,0 +1,184 @@
+    """
+
+    if _github_installations[installation_id] is not None:
+        expires_at, oauth_token = installation_id[installation_id]
```
`installation_id[installation_id]`

LordAro commented on dorpsgek/github.py:
```diff
@@ -0,0 +1,184 @@
+
+    return jwt.encode(data,
+                      key=GITHUB_PRIVATE_KEY,
+                      algorithm="RS256").decode()
```
doesn't need to be split over several lines

LordAro commented on requirements.txt:
```diff
@@ -0,0 +1,16 @@
+aiohttp==3.3.2
```
Not that I've examined a huge amount of python repos, but I think this is unusual. Usually requirements.txt just has the basic requirements (requirements.base), but with some non-specific version requirements (`>=1.2.3, <2.0.0`, etc)

Assuming you want this file for reproduceability reasons, maybe requirements.lock or something?

TrueBrain replied:
It is the other way around. `requirements.txt` pins the exact version, and often `setup.py` supplies the range. In this case, `requirements.base`., because I was looking too much how Python did their bot :D (and because I won't have a `setup.py`).

`requirements.txt` is the exact version production will be running. So a bump to this file, will also cause a redeployment.

LordAro reviewed (changes requested):
Perhaps a setup.py as well? Could be used to specify minimum python (3.6?) at least

TrueBrain commented:
I wanted to avoid a `setup.py`, as it feels a bit of overkill. Not sure I can maintain that stance .. :D

TrueBrain committed:
Enable flake8 validation

You can test it yourself by running 'test.sh', after you installed the required packages

TrueBrain committed:
Add a Dockerfile to build and run the DorpsGek GitHub App

TrueBrain commented:
I did add a `setup.py` after all; just a bit nicer I guess.

LordAro commented on Dockerfile:
```diff
@@ -0,0 +1,24 @@
+        || ( \
+            echo "!! ERROR !! different packages installed than defined in requirements.txt:" \
+            && diff -u --strip-trailing-cr requirements.txt requirements.installed \
+            && exit 1 ) 1>&2
```
Overly complicated, `diff -uw foo bar` is all that's required - no need for test

Could also combine the diffs - `diff -wu requirements.txt requirements.installed || (echo "Error..." && exit 1)`

LordAro replied:
Though I wonder why the test is even necessary - don't feel like we're expected to be checking pip for bugs

TrueBrain replied:
We are not checking pip for bugs. We are checking if `requirements.txt` is installed as suggested. Not that someone fucked up `requirements.txt`. So we are checking the developer :)

LordAro commented on Dockerfile:
```diff
@@ -0,0 +1,24 @@
+COPY dorpsgek_github /bot/dorpsgek_github
+
+RUN pip install -r requirements.txt
+RUN pip install -e .
```
why -e step?

TrueBrain replied:
To install the local package too. This makes things like `entry_points` available.

LordAro commented on dorpsgek_github/__main__.py:
```diff
@@ -0,0 +1,45 @@
+
+
+if __name__ == '__main__':
+    main(sys.argv[1:])
```
Don't see any indication of what `sys.argv[1:]` is supposed to be? is it optional?

TrueBrain replied:
Common practice; it just passes the argv from command line through, like entry_point does.

TrueBrain replied:
(and currently we don't do anything with the command-line; but the entry_point forces us to have it as parameter to `main()`, so we might as well pass it along if you call this module with `python -m dorpsgek_github`.

TrueBrain replied:
Meh; I remembered the wrong way around, and `entry_point` does NOT add it. So I removed it completely :D

LordAro commented on dorpsgek_github/github.py:
```diff
@@ -0,0 +1,182 @@
+_github_installations = {}
+
+
+# HACK - Monkeypatch a fix that is already upstream
```
link to commit that fixes this? Or an expected release number? Similar for the other "HACK"s

LordAro commented on dorpsgek_github/github.py:
```diff
@@ -0,0 +1,182 @@
+
+    event = sansio.Event.from_http(headers,
+                                   data,
+                                   secret=GITHUB_APP_SECRET)
```
it's a stylistic choice, but I don't like it when parameters are unnecessarily split over several lines

TrueBrain replied:
either split all parameters, or none. So if it doesn't fit on one line, give them all their own line. Something like:
```
headers, data,
secret=GITHUB_APP_SECRET
```
is spacial very hard to read.

TrueBrain replied:
In this case it fits on a single line again, so that is an easy fix :P

LordAro commented on dorpsgek_github/modules/registration.py:
```diff
@@ -0,0 +1,25 @@
+
+@github.register("installation_repositories", action="added")
+async def installation_repositories_added(event, github_api):
+    pass
```
intentionally left blank, or todo? would appreciate a comment

LordAro commented on .travis.yml:
```diff
@@ -0,0 +1,13 @@
+python:
+  - "3.6"
+before_install:
+  - python3 -m pip install -U flake8
```
any particular reason for splitting this into before_install?

TrueBrain replied:
How travis likes it :)

LordAro commented on dorpsgek_github/__main__.py:
```diff
@@ -0,0 +1,45 @@
+from dorpsgek_github.config import GITHUB_APP_PORT
+
+# List of modules that are enabled
+from dorpsgek_github.modules import (  # noqa
```
why is this `#noqa` necessary?

TrueBrain replied:
`flake8` otherwise rightfully says the imports are not used. He doesn't know there is a registry inside those files that are now kicked into working. So we know the imports are not used, and we are fine with that :) so we mute `flake8` from complaining

LordAro reviewed (commented):
Also have a general check of which quote style to use - it's inconsistent in several places - maybe flake can also check that?

TrueBrain commented:
The coding style is validated via `flake8`. So please point out what you consider inconsistent :) Both `flake8` as `pycharm` don't notice any.

TrueBrain commented:
Removed the `setup.py` again; having a version number in there is confusing, as we will be running HEAD in production; bumping the version for every PR is silly, so having no version is better :D
